### DISCUSSION TITLE
Do without while clause

### DISCUSSION THREAD

**Discussion by carnitron** [dotnet/csharplang#6436] [Language Ideas]
Minor feature request that has bugged me for decades.  Sorry if this has been discussed already in the past.

I'd like to be able to have a do-while statement without the while clause.  Basically just a do.  Heck, we don't even need a do, it could just be a code block.

The whole point is to use the break to break out of the block when a test expression fails, since this is the clearest, most elegant way to write code with complex compound tests.

You see code like this:

```
// Some crazy compound expression that is a lot more terrible to visually parse than my
// simplifying letters here suggest.
if (A && B && (C || D) && (!E && !F && G))
{
   DoSomething();
}
```

Or code like this:

```
// I've literally seen these go off the right edge of the screen in some uglier code bases.
if (A)
{
    if (B)
    {
        if (C || D)
        {
            if (!E && !F && G)
            {
                DoSomething();
            }
        }
    }
}
```

I use do-while statements like this:

```
do
{
    if (!A)
        break;

    if (!B)
        break;

    if (!C && !D)
        break;

    if (E || F || !G)
        break;

    DoSomething();
}
while(false)
```

The advantage here is that each test is split out into an easy-to-digest piece that reads like a simple to-do list.

This works great, but the "while(false)" is weird and unnecessary in this context.

Yes, you can use goto statements, but this runs the risk of annoying religious battles.

I just want to do this:

```
do
{
    if (!test1)
        break;

    if (!test2)
       break;

    DoSomething();
}
```

Or even this:

```
{
    if (!test1)
        break;

    if (!test2)
       break;

    DoSomething();
}
```

Basically this is a very common case for me and I want to type less, particularly since whatever extra stuff I have to type seems unnecessary and clarity-reducing.

Thoughts?  Thanks for reading.

**Comment by HaloFour** [comment]
Sounds like a good use case for a local function.  Or you could just use `goto`.

**Reply by TonyValenti** [reply]
I think what you want is a for loop that isn't incrementor-based.

**Reply by HaloFour** [reply]
From how I'm interpreting the proposal the goal does not seem to be to loop but to execute the block exactly once, with the ability to jump to the end of the block via a `break` statement.  The desire is to emulate `goto` without using `goto`:

> Yes, you can use goto statements, but this runs the risk of annoying religious battles.

It's not the position of the C# language or the design team to design new features around stylistic preferences, regardless of how religious they get.  The `goto` statement exists in C# for a reason, to handle these specific use cases.  The team is exceedingly unlikely to consider adding a second `goto`-except-in-name if only to placate those stylistic preferences.

**Reply by carnitron** [reply]
It is my impression that C# has frequently added "syntactic sugar" features expressly to make code simpler and cleaner to write and read, which is hopefully the point of style for most people.

It's a fair point that the simplicity improvement over the goto is small, though I comment below on my issue with the goto.

**Reply by HaloFour** [reply]
Simpler and cleaner, yes.  Not to specifically avoid distaste for existing features that perform the exact functionality.

**Reply by CyrusNajmabadi** [reply]
> It is my impression that C# has frequently added "syntactic sugar" features expressly to make code simpler and cleaner to write and read, which is hopefully the point of style for most people.

I think the issue here is that there does appear to be a very simple and clean option available. But it's not being used because of "stigma". That's a local policy decision that isn't a suitable reason for us to build something new. 

Let's put it another way. You have a team that wants this new feature because they have stigmatized 'goto'.  Ok. So say we now add the feature you want. Now your team (or some other team) stigmatizes this new feature *as well*. What then?  Do we now make a third feature for the same purpose? What if your team (or others) stigmatize that? Etc. Etc.

**Reply by TahirAhmadov** [reply]
Yes, this is definitely a situation which calls for `goto`.

**Reply by jnm2** [reply]
I like `goto` a lot, but I don't like having to invent a label name just because my loop doesn't have a condition.

**Reply by foxesknow** [reply]
> Simpler and cleaner, yes. Not to specifically avoid distaste for existing features that perform the exact functionality.

Really? Isn't that what expression bodied members and the using statement do, providing no new functionality?

**Reply by HaloFour** [reply]
> Really? Isn't that what expression bodied members and the using statement do, providing no new functionality?

Those were added to provide consistency with the syntax used for lambdas.  More importantly, they were added because the language team felt they were worth adding, and they are the final arbiters of the language.

**Reply by foxesknow** [reply]
> > Really? Isn't that what expression bodied members and the using statement do, providing no new functionality?
> 
> More importantly, they were added because the language team felt they were worth adding, and they are the final arbiters of the language.

I'm not sure how a using statement provides consistency with the lambda syntax... 

But your statement:

> Not to specifically avoid distaste for existing features that perform the exact functionality

Is clearly false.

**Reply by HaloFour** [reply]
@foxesknow 

> I'm not sure how a using statement provides consistency with the lambda syntax...

Expression-bodied members, not using statements.  For using statements it was demonstrated that the scope of the resource more often than not equals that of the scope of the method and requiring a level of indentation per resource type was a wasteful use of real estate.

> But your statement:
>
> > Not to specifically avoid distaste for existing features that perform the exact functionality
>
> Is clearly false.

Neither of those features were added because people claimed that their teams forbid the use of the existing features.  Please stop inventing strawmen arguments.  There are thousands of proposals on this repository and a Cartesian hypercube of opinions regarding which are more important or valuable than others.  At the end of the the day the language designers are _the final_ arbiters as to what features are to be adopted into the language, and they are not required to justify their decisions against your personal preferences.

**Comment by TonyValenti** [comment]
Try something like this:
````
for(;  test1 && test2   ;){
  DoSomething();
}
````

**Comment by TonyValenti** [comment]
Also, for some loops, I just do this:
```
while(true) {
   DoSomething();

  if(condition){
    break;
  }
}

**Comment by TonyValenti** [comment]
Also, one last thing that I'll do sometimes is something like this:
```

var ShouldExecute = false
   || (Condition1)
   || (Condition2)
   ;
if(ShouldExecute){
  DoSomething();
}


```

**Comment by carnitron** [comment]
Thx for the suggestions guys.  I updated the text in my original post to point out that I'm trying to handle this situation with less typing.  I think it it's clearly possible to express what needs to be handled here with less text/syntax, and that in doing so, we actually increase clarity.  I'm trying to avoid extra or unusual language syntax to express this basic case clearly.

**Reply by CyrusNajmabadi** [reply]
This feels like exactly what a goto is for :-)

**Reply by carnitron** [reply]
So the problem with the goto solve is that -- for whatever historical/cultural reasons -- gotos have serious stigma attached to them.

I've been in all sorts of code bases over the 30+ years I've been a software developer, from ugly to super sophisticated, and I honestly cannot remember the last time I've seen one used.  The cultural bias against them is just too strong.

Using the do-while construct has always been easier and "safer", so I've ultimately pulled out of any goto solves before checking them in.  Software development is hard enough, without being that guy that checks in code everyone hates, lol.

We can legitimately debate whether this is fair or reasonable, but it will likely have zero effect on the reality software developers are faced with here.

**Reply by CyrusNajmabadi** [reply]
> gotos have serious stigma attached to them.

This seems to be an issue with the team you work on. That seems much easier to address than creating a new language feature with a name like `goto2` to appease them :-)

You're welcome to have your coworkers come here to discuss things if you'd like. We don't believe there's us anything wrong with `goto`, and we'd be fine explaining that to them.

**Reply by HaloFour** [reply]
That stigma is seriously outdated and doesn't apply to many modern programming languages that offer the keyword.  Languages like C# ensure that you cannot jump between contexts in a way that would corrupt state, for instance.  That is why C# continues to offer the keyword and suggests using it in those cases where it is appropriate.

IMO the use of `do`/`while` in these cases is a lot worse than `goto` as it implies a loop but you're not using it for a loop, you're using it specifically so that you can use `break` like `goto` without having the name `goto`.

There are a lot of stylistic preferences that will come into play here.  I'd likely ding you for your choice of using a loop.  Others dislike the idea of "early returns" and would likely ding you for using `break` as a way to achieve it.  C# offers enough different ways to accomplish these same goals that I don't think we need another form of "goto" in the language.

**Reply by CyrusNajmabadi** [reply]
>  Software development is hard enough, without being that guy that checks in code everyone hates, lol.

Everything we build is hated by some group. (seriously, just see the backlash we get on almost everything we make). We can't produce duplicate features to placate these groups as it will just be an unending path of making something, having people go "nope, I hate it, do it again" over and over again. 

If we have a feature that solves the problem and some group intentionally chooses not to use it. That's on them. :-)

**Reply by carnitron** [reply]
> This seems to be an issue with the team you work on. That seems much easier to address than creating a new language feature with a name like goto2 to appease them :-)
> 
> You're welcome to have your coworkers come here to discuss things if you'd like. We don't believe there's us anything wrong with goto, and we'd be fine explaining that to them.

It's not just the team I'm working on now, it's all the teams I've worked on.

I appreciate your offer to explain gotos to people, but you have to understand the optics of this.

At my last company, coding standards were typically heavily debated by engineers and technical leadership during team forming, with input from the team itself and with technical leadership from other teams.  And then once decisions were finalized, we put these discussions largely to bed, so that people could focus on getting work done.  Some of these people were among the industry elite.

So now I'm supposed to say:  "hey guys, I know we spent all this time getting consensus on how we are going to use the language, but some random guy on the internet wants to explain to us how gotos work and why we're all wrong."

That *would* actually be amusing, but I hope you can understand why it's not happening, lol.

Also, just to be clear, I'm not suggesting we add goto2.  I'm suggesting we remove extraneous syntax from a feature that is already in the language.

In any event, are you and @HaloFour community people that work with Microsoft to filter out user requests?  This is my first post here, so I don't have a sense of the community.

**Reply by CyrusNajmabadi** [reply]
> but some random guy on the internet wants to explain to us how gotos work and why we're all wrong."


I'm confused by this. You think they would accept a new feature by me, but they wouldn't accept a discussion with be on an existing feature?

Also, the above doesn't address at all what happens when that same group rejects the new feature and again refuses to discuss it. What then? We make a third goto form? :-)

**Reply by 333fred** [reply]
HaloFour is a community member (who we greatly appreciate). Cyrus and myself are members of the language design team, and it's us you need to convince if you want your feature to move forward.

> So now I'm supposed to say: "hey guys, I know we spent all this time getting consensus on how we are going to use the language, but some random guy on the internet wants to explain to us how gotos work and why we're all wrong."
>
> That would actually be amusing, but I hope you can understand why it's not happening, lol.

Frankly, that sounds a heck of a lot easier than implementing a language change. Even small features require hundreds of hours of work, between specification, implementation in the compiler, tooling work, documentation, user training, etc. That's a hard pill to swallow when goto already exists for this purpose.

**Reply by carnitron** [reply]
Ok, thanks for the explanation @333fred.  When you put it that way, I can understand the hesitance.  Cheers.

**Reply by HaloFour** [reply]
There are also _many_ different ways to accomplish the same thing using other parts of the C# language.  You mention a few in your proposal, such as nested/reordered `if` statements.  You may have some distaste in writing code like that but it is idiomatic and can be just as clear depending on how you structure the statements and name things.  "Early returns" is something that also raises hackles in the overly opinionated and you'll find just as many teams that find the use of `break`/`return` here distasteful.

**Comment by iam3yal** [comment]
Another option is to use the language to create some sort of a wrapper or DSL to achieve this something like this:
```csharp
new Condition()
    .If(A)
    .And(B)
    .And(C || D)
    .And(!E && !F && G)
    .Then(() => DoSomething())
```

**Reply by FaustVX** [reply]
IMO, the problem with your proposal is that `A`, `B`, `C && D` and `E || F || G` will be evaluated.
So if `A` is `value is not null` and `B` is `value.CanDoSomething()` then you will get an exception.

**Reply by iam3yal** [reply]
@FaustVX I didn't provide any implementation it's merely APIs so getting an exception has nothing to do in particular with my suggestion but if you get an exception you will also get an exception with any of the above implementations so there is no problem here, really.

**Reply by jnm2** [reply]
@FaustVX Easy to work around with lambdas.

**Reply by FaustVX** [reply]
@jnm2 Yes, with lambdas, this could work, but you'll still loose nullable flow analysis.
@eyalalonn
```C#
object? value = null;
new Condition()
    .If(value is not null)
    .And(value.CanDoSomething) // <- The runtime will throw an exception here, trying to access value which is null
    .Then(() => DoSomething())
```

[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRII5wJZICaoC2EhwECyANDCAnAHaV4gDUAPgMT1wA2PUwHhAAEEegKEBYAFAyAAgCYAjDJkA1KDzgQA/MIBum7cIC8w7nwDcq6fQgB3YQGEA9vTzYY2NwAoAlDLCQcIAdACSYD6GWiLYyOYuMOa8PAHSwaEAgu5RRhAhTlD0ACIuAMouxDAAFtj0AOZ+wgD0zc5lABwAbAAMCiDCxeQQkEj0AMYiLmDCUMIADi7IyNiCAJ7JfMJIo2KTIcKZAHLFwjUILo5HKQBKI8MTEACiAB6T815ugcEhACrVYlyMRCpQqVVqDT81lk0jkAGZhIpnG4PJ96MIAN7fILw5HuTzedERHzAFwuHgGPJpDI04ImAB8Z1qyGhGVxrnxaMOOVJ5MpMWptJpDKZcVZwVxcgALMJ/oC5EoFLNxmjBbSRVAVYT/NCAL42SVKjQxGRY9IShG8imFErlSoQGp1eqmRlgTTICDinEI6WDO3gp3+TH66S6oA==)

**Reply by TahirAhmadov** [reply]
In addition to the practical reasons outlined by @FaustVX, in general, there is almost never a need to re-implement existing language constructs as methods.

**Reply by iam3yal** [reply]
@FaustVX Yes I get the point that the evaluation doesn't short circuit but as @jnm2 you can use a lambdas to workaround it.

@TahirAhmadov In general I agree with you but It's not about reinventing the wheel for the sake of it but creating abstractions that address actual occurring problems in the code and when the language or the platform as a whole doesn't support/provide exactly what you need I see nothing wrong with people creating abstractions to get exactly or almost what they need to get the job done, sometimes with some sacrifices, also, my example is incomplete and was merely an idea I don't expect people to use it as is.

**Comment by ds5678** [comment]
I think this is a reasonable solution to your problem.
```cs
if (!A)
{
}
else if (!B)
{
}
else if (!C && !D)
{
}
else if (E || F || !G)
{
}
else
{
	DoSomething();
}
```
It's about the same amount of lines and doesn't require changes to the language.

**Comment by TahirAhmadov** [comment]
This got me thinking about another aspect of the developer experience - step by step debugging. If all those conditions were inverted and combined in a single `if` statement whose "then" block would call `DoSomething`, it would be the cleanest way of expressing the intent. The conditions can be spread over multiple lines and indented, as needed, to preserve readability.
```cs
// Some crazy compound expression that is a lot more terrible to visually parse than my
// simplifying letters here suggest.
if (
  A && // some long condition spanning almost entire width of a typical display
  B && // same as above
  (C || D) && // same
  (!E && !F && G) // same
)
{
   DoSomething();
}
```
This leaves the question of, how is it worse than multiple `if` statements with `goto`s? Answer - it's more difficult to debug. There is no easy way to check the result of each condition easily - the only way would be to paste each part of the big `if` condition to a watch window, which is cumbersome. This happened to me a few times, as well.

Does it mean we should make all expressions compiled in a way that allows "stepping over" each operator? Definitely not. However, an argument can be made that either 1) "step into" command should be able to do this; or 2) there should be a new command, "step through expressions", which would accomplish this. 
Obviously, both of these solutions require extensive work on the compilation and tooling side, and is probably not worth it in the grand scheme of things. Also, at that point, it's no longer a language design question. But I figured I'd bring it up anyway - in case this is the pain point of the OP author.

**Reply by HaloFour** [reply]
Tooling is definitely able to capture and display the result of expressions, like how Visual Studio can show `$ReturnValue` or `$Exception` pseudo-variables today.

However, if you have that much logic compacted into a single `if` statements then breaking that up over multiple statements is already good practice, and you can accomplish that combined with the "short-circuiting" manually by nesting `if` statements.  That is the idiomatic way to approach this.  If you want a fall-through behavior as suggested in this proposal, that can also be accomplished with local functions, where you can use early returns.

**Reply by TahirAhmadov** [reply]
@HaloFour I know what you mean, this is not a show stopper by any means. 

>  if you have that much logic compacted into a single if statements

That's not always the case, it's common to have a few long lines with simple statements, such as `enum` comparisons with long names, a method call with many arguments, etc. It's not a lot of logic, perhaps 3 logical conditions combined in an AND, but it is a lot of text.

What I wrote also applies to other situations, not just `if` statements; imagine a simple assignment with a somewhat complex mathematical expression on the right, even if it fits on one line. In general, it would be somewhat helpful to step through expressions (again, it's not a dire need).

**Reply by HaloFour** [reply]
@TahirAhmadov 

> imagine a simple assignment with a somewhat complex mathematical expression on the right, even if it fits on one line. In general, it would be somewhat helpful to step through expressions (again, it's not a dire need).

Sounds like a tooling question.  I'd wager the IDE can already do that based on the IL and debugging symbols but it's more a matter of how annoying it would be to step at that granularity in typical circumstances.  Even in those cases if I'm that interested in the intermediate results I will often break it up into multiple expressions which assign to locals.
